aws_signing_helper/pkcs11_signer.go

```diff
@@ -0,0 +1,214 @@
+		if err != nil {fmt.Println("Error parsing certificate")}
+		serialNumber := cert.SerialNumber.String()
+		certLocated.Subject = cert.Subject.CommonName 
+		certLocated.Issuer = cert.Issuer.CommonName
+		certLocated.SerialNumber, _ = new(big.Int).SetString(serialNumber, 16)
+		if reflect.DeepEqual(certLocated, certIdentifier) {
+			return i, cert, nil
+		} else {
+		}
+	}
+
+
+	return 1, nil, errors.New("unsupported certificate")
+}
+
+func (pkcs11Signer PKCS11Signer) Public() crypto.PublicKey {
+	return nil
+}
+
+func (pkcs11Signer PKCS11Signer) Close() {
+}
+
+func (signer PKCS11Signer) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+		lib := signer.libpkcs11
+		pin := signer.pinpkcs11
+		id := signer.idpkcs11
+
+		p := pkcs11.New(lib) 
+		err = p.Initialize()
+		if err != nil {
+			panic(err)
+		}
+
+		defer p.Destroy()
+		defer p.Finalize()
+
+		slots, err := p.GetSlotList(true)
+		if err != nil {
+			panic(err)
+		}
+
+		session, err := p.OpenSession(slots[0], pkcs11.CKF_SERIAL_SESSION|pkcs11.CKF_RW_SESSION)
+		if err != nil {
+			panic(err)
+		}
+		defer p.CloseSession(session)
+
+		err = p.Login(session, pkcs11.CKU_USER, pin)// err = p.Login(session, pkcs11.CKU_USER, pin) 
+		if err != nil {
+			panic(err)
+		}
+		defer p.Logout(session)
+
+		templatepk := []*pkcs11.Attribute{
+			pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PRIVATE_KEY),
+		}
+		_ = p.FindObjectsInit(session, templatepk) 
+		objspk, _, _ := p.FindObjects(session, 100)
+		_ = p.FindObjectsFinal(session)  // found nothing
+
+		err = p.Login(session, pkcs11.CKU_CONTEXT_SPECIFIC, pin)
+
+		err = p.SignInit(session, []*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_SHA256_RSA_PKCS, nil)}, objspk[id]) // CKM_SHA256_RSA_PKCS
+		if err != nil {
+		fmt.Println("err: ", err)
+			log.Fatalf("Signing Initiation failed (%s)\n", err.Error())
+		}
+
+		sig, err := p.Sign(session, digest)
+		if err != nil {
+		err = fmt.Errorf("Signing failed (%s)\n", err.Error())
+		fmt.Println("err: ", err)
+		}
+
+
+		if err == nil {
+			return sig, nil
+		}
+
+	log.Println("unsupported algorithm")
+	return nil, errors.New("unsupported algorithm")
+}
+
+func (pkcs11Signer PKCS11Signer) Certificate() (*x509.Certificate, error) {
+	return pkcs11Signer.cert, nil
+}
+
+func (pkcs11Signer PKCS11Signer) CertificateChain() ([]*x509.Certificate, error) {
+	return pkcs11Signer.certChain, nil
+}
+
+
+// Returns a PKCS11Signer, that signs a payload using the
+// private key passed in
+func GetPKCS11Signer(certIdentifier CertIdentifier, libpkcs11 string, pin string) (signer Signer, signingAlgorithm string, err error) {
+	idpkcs11, cert, err := GetMatchingPKCSCerts(certIdentifier, libpkcs11)
+	if err != nil {
+		return nil, "", err
+		//errors.New("No matching certificate found on the security device")
+	}
+	pinpkcs11 := pin
+    // case where there are no matching identities is already handled as an error from GetMatchingCerts
```

Comment:
Minor, but we should remove unnecessary comments (including commented code) before merging into the `main` branch.

Reply:
Done.